lib/runner/TestRunner.js

```diff
@@ -133,7 +133,47 @@ module.exports = class TestRunner {
         return testResults;
     }
 
+    async resolveVariablesForAssertionInteraction(testInteraction, resolveMethod) {
+        if (!testInteraction.assertions || !testInteraction.assertions.length) {
+            return;
+        }
+
+        // Using for without iterators to properly support async / await
+        for (let i = 0; i < testInteraction.assertions.length; i++) {
+            const assertion = testInteraction.assertions[i];
+            if (!assertion.variables || !assertion.variables.length) {
+                continue;
+            }
+            for (let j = 0; j < assertion.variables.length; j++) {
+                const variable = assertion.variables[j];
+                const variableValue = await resolveMethod(variable, testInteraction);
+                // Empty values are allowed but undefined means no replacement was found
+                if (typeof variableValue == "undefined") {
+                    return;
+                }
+                if (Array.isArray(assertion.value)) {
+                    const replacedValues = assertion.value.map((val) => {
+                        return val.split("{" + variable + "}").join(variableValue);
+                    });
+                    assertion._value = replacedValues;
+                } else {
+                    assertion._value = assertion.value.split("{" + variable + "}").join(variableValue);
+                }
+
+            }
+        }
+    }
+
+
     async batchRun(invoker, testSuite, interactions, context) {
+        // we have variables to resolve, we need to resolve them before calling invocation
+        if (testSuite.filterObject() && testSuite.filterObject().resolve) {
```

Comment:
This needs to be called AFTER the call to the virtual device. This is because the interactions with Alexa may have an affect on the state of the variables being accessed (think of a bank transfer test, where money is transferred between accounts. We want to ensure the bank balance is correct AFTER the transfer).

Reply:
@jkelvie , it makes sense, I will make the correction, but also if that is the case if the assertion has brackets, we will not send the assertions as expected phrases to avoid sending incorrect data.